### PR TITLE
ambassador-operator

### PR DESCRIPTION
**Pull request by cakuros**
## BACKGROUND
This PR adds the Ambassador Operator to the DigitalOcean marketplace.  The Ambassador Operator is a Kubernetes Operator that controls the complete lifecycle of Ambassador in your cluster. It also automates many of the repeatable tasks you have to perform for the Ambassador Edge Stack and provides a declarative method of maintaining an Ambassador Installation. Once installed, the AES Operator will automatically complete rapid installations and seamless upgrades to new versions of Ambassador.

[Ambassador Edge Stack](https://getambassador.io/) is a popular, high-performance [Ingress Controller](https://www.getambassador.io/products/edge-stack/api-gateway/) and [API Gateway](https://www.getambassador.io/learn/kubernetes-glossary/api-gateway/) built on [Envoy Proxy](https://www.envoyproxy.io/). Envoy Proxy was designed from the ground up for [cloud-native](https://www.getambassador.io/learn/kubernetes-glossary/cloud-native/) applications. Ambassador exposes Envoy's functionality as [Custom Resource Definitions](https://www.getambassador.io/learn/kubernetes-glossary/custom-resource-definition/), with integrated support for rate limiting, authentication, load balancing, observability, and more.

-----------------------------------------------------------------------

## Changes
* Adds the Ambassador Operator and README instructions for installing Ambassador Edge Stack and setting up an example mapping.

-----------------------------------------------------------------------

## Checklist
- [x ] review the [contributing doc](https://github.com/digitalocean/marketplace-kubernetes/blob/master/CONTRIBUTING.md) with steps for both adding or updating your application (if applicable)
------------------------------------------------------------------------

Reviewer: @marketplace-eng

### CODE REVIEW

**Comment by paulanunda** [comment]
Hi @cakuros, thanks for the PR. Would it be possible to install the operator via Helm 3 instead of applying YAML? This is what I tried locally, but I'm getting a `CrashLoopBackOff`:
```
helm repo add datawire https://getambassador.io
helm repo update
k create ns ambassador-operator
helm install ambassador-operator datawire/ambassador-operator --namespace ambassador-operator
```

**Comment by cakuros** [comment]
Hi @paulanunda, the helm install for the Operator requires applying the CRD's with `kubectl apply -f https://github.com/datawire/ambassador-operator/releases/latest/download/ambassador-operator-crds.yaml`.  I followed the instructions @ https://github.com/digitalocean/marketplace-kubernetes/blob/master/CONTRIBUTING.md#step-2-create-your-application-stack to render the chart as yaml and included the CRD application to the `deploy.sh` and `deploy-local.sh`.  Is there something different I should do instead?

**Comment by paulanunda** [comment]
@cakuros You did the right thing, the docs need to be updated (planned on doing it next week). You now have the ability to deploy via Helm directly instead of committing YAML to the repo that ultimately gets applied to a cluster. (example: https://github.com/digitalocean/marketplace-kubernetes/blob/master/stacks/metrics-server/deploy.sh). This way, if a user wants to update their helm-installed app, they can do so from their own Helm CLI.

I installed the CRD into my test cluster, but was still unable to get the Helm-based operator to install. I was able to get Ambassador itself installed via Helm on a different cluster. Maybe it makes sense to list that in the Marketplace?

**Comment by cakuros** [comment]
@paulanuda Ok, after some review and in light of the information you've given us, I think the best route forward would be for us to do the Helm3 install for Ambassador instead of the Operator.  Our original thinking was that version control would be simpler to do via the operator (since it updates less frequently than Ambassador and always points to the latest Ambassador version), but if you are pulling directly from helm charts, then I think that is the best way forward.  I will re-adjust the PR to reflect that and get back to you.  Thanks for the feedback!

**Comment by paulanunda** [comment]
@cakuros I've updated the https://github.com/digitalocean/marketplace-kubernetes/blob/master/CONTRIBUTING.md w/ detailed instructions on how to set up your app to deploy it w/ Helm3. If you follow those, you'll have the files you'll need for this PR (just the stacks dir).

**Comment by cakuros** [comment]
@paulanunda Awesome!  I just updated everything to use a helm install of Ambassador Edge Stack instead of the operator, and I also tested deploy.sh on both minikube and on a DO k8s cluster, both of which appear to have worked.  For the sake of maintaining continuity of this conversation/review process I will just push to this branch instead of creating a new PR if that's OK.  The formal name of the app will be 'Ambassador Edge Stack' though, not 'Ambassador Operator', just to be clear.

File: stacks/ambassador/deploy.sh [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+fi
+
+kubectl apply -f https://getambassador.io/yaml/aes-crds.yaml
+
+helm upgrade "$STACK" "$CHART" \
+  --install \
+  --create-namespace \
+  --namespace "$NAMESPACE" \
+  --values "$values" \
+  --wait
```
Comment (paulanunda) [inline stacks/ambassador/deploy.sh:34]: Did you not want to specify a Helm chart version? Doing so guarantees that any breaking upgrades don't inadvertently break installs for users.

File: utils/stack-templates/deploy.sh [hunk 2]
```diff
@@ -5,16 +5,15 @@ set -e
 ################################################################################
 # repo
 ################################################################################
-helm repo add {{HELM_REPO_NAME}} {{HELM_REPO_URL}}
+helm repo add datawire https://getambassador.io
```
Comment (paulanunda) [inline utils/stack-templates/deploy.sh:8]: I don't think you meant to modify this file :)

Comment (cakuros) [reply utils/stack-templates/deploy.sh]: Good catch ;) I was wondering why my changes didn't take effect the first time.

File: stacks/ambassador/deploy.sh (hunk 1 above)
Comment (cakuros) [reply stacks/ambassador/deploy.sh:36]: Fixed.  I also added some instructions to the docs for upgrading.

**Comment by cakuros** [comment]
Changes made!

**Pull request merged**